harness/determined/pytorch/_callback.py

```diff
@@ -91,12 +91,30 @@ def on_training_start(self) -> None:
         """
         pass
 
-    def on_training_epoch_start(self) -> None:
+    def on_training_batch_start(self, batch_idx: int) -> None:
+        """
+        Run on start of a new training batch
+        """
+        pass
+
+    def on_training_batch_end(self, batch_idx: int) -> None:
+        """
+        Run on end of a training batch
+        """
+        pass
+
+    def on_training_epoch_start(self, epoch_idx: int) -> None:
         """
         Run on start of a new training epoch
         """
         pass
 
+    def on_training_epoch_end(self, epoch_idx: int) -> None:
```

Comment:
usually `on_thing_end()` would include the metrics for that thing.  We can't actually deliver end-of-epoch metrics, due to how training windows over which metrics are calculated in our system do not align with epoch boundaries.

I'd like to fix that in the future, with the Trainer API.  But it's going to be tricky since it will involve reconsidering the sharding model we currently have.  My first thought was to add a `metrics={}` argument here that will always be empty until we support real epoch-based training.  That's probably a terrible idea though... probably another inspection hack in the future is better.  What do you think @liamcli?

Reply:
For this particular case, I would either not pass in metrics or pass in metrics aggregated over batches in an epoch (with the understanding that the last batch will not align with epoch boundary but it would at least match what we print for that training workload).  I prefer the latter since a common use case will probably be early stopping decisions based on metrics.

Reply:
the way that loop is structured, at the time when we detect an end of epoch (by seeing that we processed the number of batches that's a factor of the epoch length) we have a list of per-batch metrics for all the batches encountered so far _on this worker_. Today, we don't build an aggregate here. We wait until the end of the workload, and then aggregate metrics across batches and workers. I am guessing that is what "_way_ easier" is referring to?

Reply:
given the complications, i'm okay with just passing `epoch_idx` to this callback and pointing users to something like `on_training_workload_end` for early stopping decisions.

Reply:
> at the time when we detect an end of epoch ... we have a list of per-batch metrics for all the batches encountered so far _on this worker_

I mean, we only have the metrics we've gathered _on this worker_ since the last training workload ended, right?  I think we flush the batch metrics after every workload.

Reply:
correct: https://github.com/determined-ai/determined/blob/36bb4d9beac348e32b42cecd9da7b26956f889c1/harness/determined/pytorch/_pytorch_trial.py#L272